### PR TITLE
Component/text area

### LINKED ISSUE #12: TextArea
Create TextArea component.

Screenshot 2021-01-09 at 10 22 54

View the component in [Figma](https://www.figma.com/file/gGwSe7ilZCUmAIaJyUc4iY/Bristol-Tech-Fair-Design-System?node-id=489%3A44).

### PR DESCRIPTION
**Pull request by catherinehoward**
Closes #12

This pull request adds a new text area component.

### CODE REVIEW

**Comment by catherinehoward** [comment]
Hi Rob, I can't get my css styles to render for some reason. 

I see there is a desktop and a mobile version of the textarea. Presumably I'll need to create two different stories and the width and height will need to be passed as props?

I'm also having conflicts with my TextButton files. My dev environment is out of sync with the btf-frontend repo. I recently retrieved changes from the btf-frontend repo because I noticed you'd made some changes to the Storybook instructions. Could be why. What's the best way of fixing this? Thanks

**Comment by RobTF9** [comment]
You need to merge develop into `component/TextArea`, and sort out the merge conflicts. Always make sure you're branching off the most recent version of develop, otherwise you'll have old code and run in to conflicts.

Get that sorted out first and I'll take a look at the changes 👍

**Comment by RobTF9** [comment]
Did you create this branch before you finished working on the TextButton?

**Comment by catherinehoward** [comment]
> Did you create this branch before you finished working on the TextButton?

Possibly although not entirely sure!

**Comment by catherinehoward** [comment]
I thought I resolved the conflicts and now the Node CI/build is failing. Not sure how to fix this.

I've checked the status of my TextButton and TextArea branches and they are up to date with the remote GitHub repo.
My develop branch status is saying **'Your branch is ahead of 'origin/develop'** by 3 commits'.  Not sure I understand what's going on or what I need to do!

**Comment by RobTF9** [comment]
To fix CI, you just need to run `yarn lint:fix` in the terminal, commit the changes and then push the code up.

If the code on Github is up to date then the easiest thing to do is delete your local version, reclone it and install the packages again 👍

**Comment by catherinehoward** [comment]
I tried running yarn lint --fix, committing changes and pushing changes but get this error:
 error: GH006: Protected branch update failed for refs/heads/develop

Will try deleting my local version. Presumably you mean delete the entire btf-frontend folder? Once I've cloned the repo will I need to recreate my TextArea component files?

File: src/components/TextArea/TextArea.jsx [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+  );
+};
+
+TextArea.propTypes = {
+  id: PropTypes.string,
+  name: PropTypes.string,
+  labelText: PropTypes.string,
+  placeholder: PropTypes.string,
+  value: PropTypes.string,
+  onChange: PropTypes.func
```
Comment (RobTF9) [inline src/components/TextArea/TextArea.jsx]: These are all require fields for a TextArea

File: src/components/TextArea/TextArea.stories.jsx [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+
+export const TextAreaDefaultView = () => (
+  <TextArea
+    id="textarea1"
+    name="textarea"
+    labelText="Label"
+    placeholder="Input text"
+    value="Input"
+  />
+);
```
Comment (RobTF9) [inline src/components/TextArea/TextArea.stories.jsx:17]: Add a version that has loads of text.

File: src/components/TextArea/TextArea.styles.js [hunk 3]
```diff
@@ -0,0 +1,11 @@
...
+
+export const TextAreaWrapper = styled.div`
+  textarea {
+    display: block;
+    background: #f2f2f2;
+    border: 1px solid var(--lightblue);
+    box-sizing: border-box;
+    border-radius: 0.5em;
+  }
+`;
```
Comment (RobTF9) [inline src/components/TextArea/TextArea.styles.js]: background needs to use --lightgrey.

Layout wise chat to @Tumekab as these consistent. Are there anyways we can make the styles here re-useable?

File: src/components/TextButton/TextButton.styles.js [hunk 4]
```diff
@@ -0,0 +1,6 @@
+import styled from 'styled-components';
+
+export const TextButtonWrapper = styled.div`
+  padding: var(--small-space);
+  fontcolor: var(--lightblue);
+`;
```
Comment (RobTF9) [inline src/components/TextButton/TextButton.styles.js]: Delete these changes, not part of this pull request.

File: .vscode/launch.json [hunk 5]
```diff
@@ -0,0 +1,15 @@
...
+    "configurations": [
+        {
+            "type": "pwa-chrome",
+            "request": "launch",
+            "name": "Launch Chrome against localhost",
+            "url": "http://localhost:8080",
+            "webRoot": "${workspaceFolder}"
+        }
+    ]
+}
```
Comment (RobTF9) [inline .vscode/launch.json]: Remove using .gitignore

File: src/components/TextButton/TextButton.styles.js (hunk 4 above)
Comment (catherinehoward) [reply src/components/TextButton/TextButton.styles.js]: Not sure how these files have changed. I've not been working on them. What's the best way to delete the changes? Do I need to use the git checkout command e.g:
git checkout develop -- src/components/TextButton/TextButton.styles.js

File: src/components/TextArea/TextArea.test.jsx [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+      />
+    );
+    screen.debug();
+    fireEvent.change(screen.getByRole('textbox'), {
+      target: { value: 'modified text' }
+    });
+    screen.debug();
+    expect(handleChange).toHaveBeenCalledTimes(1);
+  });
+});
```
Comment (catherinehoward) [inline src/components/TextArea/TextArea.test.jsx:37]: Not sure I need all these tests at this stage? I'm not sure If the last test is right. I was trying to simulate modifying the text so that handleChange fires. Screen.debug() doesn't show that the text has been changed.

File: src/components/TextButton/TextButton.styles.js (hunk 4 above)
Comment (RobTF9) [reply src/components/TextButton/TextButton.styles.js]: Just delete the files from your branch 👍

File: src/components/TextArea/TextArea.test.jsx (hunk 6 above)
Comment (RobTF9) [reply src/components/TextArea/TextArea.test.jsx:37]: You can get rid of all the ones that are testing the rendering, the onChange function is the important thing to test. You can clean this up significantly by creating a stateful mock component like so:
```
import React, { useState } from 'react';
import { render, fireEvent } from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';
import TextArea from './TextArea';

const MockInput = () => {
  const [value, setValue] = useState('');

  const handleChange = (event) => {
    setValue(event.target.value);
  };

  return (
    <TextArea
      name="test"
      id="test"
      value={value}
      labelText="Label"
      onChange={handleChange}
    />
  );
};

const setup = () => {
  const utils = render(<MockInput />);
  const input = utils.getByLabelText('Label');
  return {
    input,
    ...utils
  };
};

test('It should change value when new input received', () => {
  const { input } = setup();
  fireEvent.change(input, { target: { value: 'Hey Cath!' } });
  expect(input.value).toBe('Hey Cath!');
});
```

The example of this is [here](https://testing-library.com/docs/example-input-event/).

@bristol-tech-fair/webdev-101 I'm tagging you all so you can see how to test change handlers.

The key thing to remember is that test files are just JS, so you put whatever you want in there. I've basically built a fake component with state and am then using that to test the onChange functionality.

File: src/components/TextArea/TextArea.styles.js [hunk 7]
```diff
@@ -0,0 +1,11 @@
...
+
+export const TextAreaWrapper = styled.div`
+  textarea {
+    display: block;
+    background: var(--lightgrey);
+    border: 1px solid var(--lightblue);
+    box-sizing: border-box;
+    border-radius: 0.5rem;
+  }
+`;
```
Comment (RobTF9) [inline src/components/TextArea/TextArea.styles.js]: Style wise, can you replicate what @Tumekab has done in the TextInput please.

**Review by RobTF9 — CHANGES_REQUESTED** [review]

File: src/components/TextArea/TextArea.styles.js (hunk 7 above)
Comment (RobTF9) [reply src/components/TextArea/TextArea.styles.js]: https://github.com/bristol-tech-fair/btf-frontend/pull/37/files

File: .vscode/launch.json (hunk 5 above)
Comment (RobTF9) [reply .vscode/launch.json]: Still need to delete this file.

**Review by RobTF9 — CHANGES_REQUESTED** [review]

**Comment by catherinehoward** [comment]
The files should be deleted now and I've made the required changes. If I've missed anything let me know.

**Review by RobTF9 — APPROVED** [review]
Nice work, let's get this merged! 👍
